### PR TITLE
[WIP] PreStartContainer and PostStopContainer events for isolator.

### PR DESCRIPTION
**Pull request by ConnorDoyle**
Logs from new "noop" isolator including container pre-start and post-stop events.

```
I0419 00:16:44.861812    3243 noop_isolator.go:24] Starting ...
I0419 00:16:44.862019    3243 noop_isolator.go:12] noopIsolator[Init]
I0419 00:20:38.761339    3243 noop_isolator.go:21] noopIsolator[PreStartPod]:
pod: &Pod{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:busy,GenerateName:,Namespace:default,SelfLink:/api/v1/namespaces/default/pods/busy,UID:b0492f3a-24d0-11e7-bef9-c85b76944bc7,ResourceVersion:719,Generation:0,CreationTimestamp:2017-04-19 00:20:38 -0700 PDT,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{kubernetes.io/config.seen: 2017-04-19T00:20:38.753435097-07:00,kubernetes.io/config.source: api,},OwnerReferences:[],Finalizers:[],ClusterName:,},Spec:PodSpec{Volumes:[{default-token-fl25t {nil nil nil nil nil SecretVolumeSource{SecretName:default-token-fl25t,Items:[],DefaultMode:*420,Optional:nil,} nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil nil}}],Containers:[{busy busybox [/bin/tail] [-f /dev/null]  [] [] [] {map[] map[cpu:{{100 -3} {<nil>} 100m DecimalSI}]} [{default-token-fl25t true /var/run/secrets/kubernetes.io/serviceaccount }] nil nil nil /dev/termination-log File Always nil false false false}],RestartPolicy:Always,TerminationGracePeriodSeconds:*30,ActiveDeadlineSeconds:nil,DNSPolicy:ClusterFirst,NodeSelector:map[string]string{},ServiceAccountName:default,DeprecatedServiceAccount:default,NodeName:127.0.0.1,HostNetwork:false,HostPID:false,HostIPC:false,SecurityContext:&PodSecurityContext{SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,SupplementalGroups:[],FSGroup:nil,},ImagePullSecrets:[],Hostname:,Subdomain:,Affinity:nil,SchedulerName:default-scheduler,InitContainers:[],AutomountServiceAccountToken:nil,Tolerations:[{node.alpha.kubernetes.io/notReady Exists  NoExecute 0xc4201b10d0} {node.alpha.kubernetes.io/unreachable Exists  NoExecute 0xc4201b10f0}],},Status:PodStatus{Phase:Pending,Conditions:[{PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2017-04-19 00:20:38 -0700 PDT  }],Message:,Reason:,HostIP:,PodIP:,StartTime:<nil>,ContainerStatuses:[],QOSClass:Burstable,InitContainerStatuses:[],},}
cgroupInfo: &CgroupInfo{Kind:POD,Path:/kubepods/burstable/podb0492f3a-24d0-11e7-bef9-c85b76944bc7,}
I0419 00:20:40.444813    3243 noop_isolator.go:31] noopIsolator[PreStartContainer]:
podName: busy
containerName: busy
I0419 00:21:25.066360    3243 noop_isolator.go:36] noopIsolator[PostStopContainer]:
podName: busy
containerName: busy
I0419 00:21:27.196435    3243 noop_isolator.go:26] noopIsolator[PostStopPod]:
podName: 
containerName: 
cgroupInfo: &CgroupInfo{Kind:POD,Path:/kubepods/burstable/podb0492f3a-24d0-11e7-bef9-c85b76944bc7,}
```

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
Added a no-op (logging) isolator for demonstration purposes.

File: pkg/kubelet/cm/event_dispatcher.go [hunk 1]
```diff
@@ -41,6 +41,14 @@ type EventDispatcher interface {
 	// PostStopPod is invoked after all of a pod's containers have permanently
 	// stopped running, but before the pod sandbox is destroyed.
 	PostStopPod(cgroupPath string) error
+
+	// PreStartContainer is invoked before the container is created.
+	PreStartContainer(podName, containerName string) (*lifecycle.EventReply, error)
```
Comment (flyingcougar) [inline pkg/kubelet/cm/event_dispatcher.go:59]: @ConnorDoyle Is podName and container name enough ?  i.e we have 2 containers in pod and we want to have 2 cpus for container "busy1" and 3 cpus for container "busy2" (separate OIR's entries in podSpec for each container). Then with `PreStartPod` call we setup limit in cpuset for podSandbox group assigning in total 5 cpus.

Then when we call `PreStartContainer` (i.e for busy1 container) we want to chose 2 cpus (from the 5 already assigned).
How will isolator know the 2 cpus should be chosen and not 3 or 4 ?

One option would be for the isolator to either save podSpec from `PreStartPod` Event, or request podSpec directly from k8s API but those approaches put more complexity on isolator.

IMO If we change `PreStartContainer` to sent container name and podSpec or containerConfig the logic in isolator will be simpler. What do you think ?

File: pkg/kubelet/kuberuntime/kuberuntime_container.go [hunk 2]
```diff
@@ -77,6 +79,26 @@ func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, podSandb
...
+	// Delegate to registered isolators.
+	eventReply, err := containermanager.GetEventDispatcherSingleton().PreStartContainer(podSandboxConfig.Metadata.Name, containerConfig.Metadata.Name)
+	if err != nil {
+		return "", fmt.Errorf("failed to collect isolator responses for container %q: %v", containerConfig.Metadata.Name, err)
+	}
+	if eventReply.Error != "" {
+		return "", fmt.Errorf("isolator returned error: %s", eventReply.Error)
+	}
+	// Apply aggregated isolator responses to the container create config.
+	for _, control := range eventReply.IsolationControls {
```
Comment (flyingcougar) [inline pkg/kubelet/kuberuntime/kuberuntime_container.go]: @ConnorDoyle I wonder if we can move that logic somewhere inside eventDispatcher interface.
In future when there will be a need to add i.e modification of env variables as in #50 the code changes will be limited hopefully only (obviously that depends on what the isolator does) to event_dispatcher.go file thus making it easier and less "impacting" on core kubelet code. Example `ResourceConfigFromReplies()` function.
What are your thoughts ?

File: pkg/kubelet/cm/event_dispatcher.go (hunk 1 above)
Comment (ConnorDoyle) [reply pkg/kubelet/cm/event_dispatcher.go:59]: Could go either way. I'll take a look at how much we'd have to change to send the pod spec in PreStartContainer. If it's too big for the PoC then I feel like it's pretty easy for the isolator to hold onto the copy it receives in PreStartPod and look it up when it receives PreStartContainer.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go (hunk 2 above)
Comment (ConnorDoyle) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: Good idea, will make a new function for it.

**Comment by squall0gd** [comment]
How about tracking CPUs with `preStartContainer` in isolator? We should change method of tracking resources due our current structure is simple and won't recognize between Pod and Container and won't make any correlation between them. So we probably need to store this information with something like this:

```go
type reservedCoresContainer struct {
   cpus []int
}

type reservedCoresPod struct {
   containers []*reservedCoresContainer  
   availableCPUs []int
   reservedCPUs []int
} 
```

**Comment by ConnorDoyle** [comment]
@squall0gd the current isolator implementation is still fine -- it just does pod-level core affinity. We can build another example to demonstrate per-container core affinity.

**Comment by squall0gd** [comment]
@ConnorDoyle why have you merged your PR with approval?

**Pull request merged**